azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -1,146 +1,122 @@
-        if requested_version.major == CURRENT_VERSION.major and requested_version.minor == CURRENT_VERSION.minor and requested_version.patch == CURRENT_VERSION.patch:
-            return AgentUpgradeType.Hotfix
-        return AgentUpgradeType.Normal
+            self._updater = RSMVersionUpdater(self._gs_id, self._daemon_version, datetime.datetime.min)
 
     @staticmethod
-    def __get_daemon_version_for_update():
+    def _get_daemon_version_for_update():
         daemon_version = get_daemon_version()
         if daemon_version != FlexibleVersion(VERSION_0):
             return daemon_version
-        # We return 0.0.0.0 if we failed to retrieve daemon version. In that case,
+        # We return 0.0.0.0 if daemon version is not specified. In that case,
         # use the min version as 2.2.53 as we started setting the daemon version starting 2.2.53.
         return FlexibleVersion("2.2.53")
 
-    def __get_next_upgrade_times(self, now):
+    @staticmethod
+    def _get_rsm_update_state_file():
+        # This file keeps if last attempted update is rsm or not.
+        return os.path.join(conf.get_lib_dir(), "rsm_update.json")
+
+    def _save_rsm_update_state(self, is_last_update_with_rsm):
         """
-        Get the next upgrade times
-        return: Next Hotfix Upgrade Time, Next Normal Upgrade Time
+        Save the rsm state to the file when we switch between rsm and self-update
         """
+        try:
+            with open(self._get_rsm_update_state_file(), "w") as file_:
+                json.dump({"isLastUpdateWithRSM": is_last_update_with_rsm}, file_)
+        except Exception as e:
+            logger.warn("Error updating the RSM state ({0}): {1}", self._get_rsm_update_state_file(), ustr(e))
 
-        def get_next_process_time(last_val, frequency):
-            return now if last_val == datetime.datetime.min else last_val + datetime.timedelta(seconds=frequency)
-
-        next_hotfix_time = get_next_process_time(self.update_state.last_attempted_hotfix_update_time,
-                                                 conf.get_hotfix_upgrade_frequency())
-        next_normal_time = get_next_process_time(self.update_state.last_attempted_normal_update_time,
-                                                 conf.get_normal_upgrade_frequency())
+    def _get_is_last_update_with_rsm(self):
+        """
+        Returns isLastUpdateWithRSM from the state file.
+        """
+        if not os.path.exists(self._get_rsm_update_state_file()):
+            return False
 
-        return next_hotfix_time, next_normal_time
+        try:
+            with open(self._get_rsm_update_state_file(), "r") as file_:
+                return json.load(file_)["isLastUpdateWithRSM"]
+        except Exception as e:
+            logger.warn(
+                "Can't retrieve the isLastUpdateWithRSM from rsm state file ({0}), will assume it False. Error: {1}",
```

Comment:
assuming False may be incorrect.

the file just contains a boolean; you can eliminate this error condition by using the presence of the file as the boolean: if it exits, it's RSM, otherwise it's sefl-update

Reply:
well, even true is not right all the time. I feel assume false and self-update more safe than true. With rsm, we need extra flags to move forward with update. If it's not there we fail the update. That will stop the update but self-update at least customers vms on latest version. Later in time switch back to RSM if flags are showed up.

As far as presence file as boolean. I already check the presence in Line #100. So at this point we can assume true if that is what we want

Reply:
Exactly, assuming True is not safe either.

Hence my suggestion of just using an empty file, and if the file exist the value (is RSM) is True,  otherwise it is False

Reply:
This is what I mean
```

def _get_is_last_update_with_rsm(self):
    return os.path.exists(self._get_rsm_update_state_file())

```
and change _save_rsm_update_stateto create the file when its argument is True and delete it when it is False.